### PR TITLE
distributed trainning for transformer

### PR DESCRIPTION
**Pull request by xuezhong**
distributed trainning for transformer,two options:
1 device cpu, which can run only with cpu instead of GPU
2 local, which can run on multi machines

### CODE REVIEW

File: fluid/neural_machine_translation/transformer/train.py [hunk 1]
```diff
@@ -257,167 +246,267 @@ def train(args):
...
+
+        optimizer = fluid.optimizer.Adam(
+            learning_rate=lr_decay,
+            beta1=TrainTaskConfig.beta1,
+            beta2=TrainTaskConfig.beta2,
+            epsilon=TrainTaskConfig.eps)
+        optimizer.minimize(sum_cost)
+
+    def train_loop(exe, train_progm):
+        def read_multiple(reader,
```
Comment (panyx0718) [inline fluid/neural_machine_translation/transformer/train.py]: It seems there is no need to move read_multiple and split_data to here?

File: fluid/neural_machine_translation/transformer/train.py [hunk 2]
```diff
@@ -257,167 +246,267 @@ def train(args):
...
+            inst_num_per_part = len(data) // num_part
+            return [
+                data[inst_num_per_part * i:inst_num_per_part * (i + 1)]
+                for i in range(num_part)
+            ]
+
+        # Initialize the parameters.
+        if TrainTaskConfig.ckpt_path:
+            fluid.io.load_persistables(exe, TrainTaskConfig.ckpt_path)
+            #lr_scheduler.current_steps = TrainTaskConfig.start_step
```
Comment (panyx0718) [inline fluid/neural_machine_translation/transformer/train.py]: why is this line commented?

File: fluid/neural_machine_translation/transformer/train.py [hunk 3]
```diff
@@ -257,167 +246,267 @@ def train(args):
...
+            TrainTaskConfig.warmup_steps)
+
+        optimizer = fluid.optimizer.Adam(
+            learning_rate=lr_decay,
+            beta1=TrainTaskConfig.beta1,
+            beta2=TrainTaskConfig.beta2,
+            epsilon=TrainTaskConfig.eps)
+        optimizer.minimize(sum_cost)
+
+    def train_loop(exe, train_progm):
```
Comment (panyx0718) [inline fluid/neural_machine_translation/transformer/train.py]: this function seems too long. Maybe it should be moved out of of train?

**Review by panyx0718 — COMMENTED** [review]
Can you simply ```train``` function? For example, there is no need to put ```read_multiple``` and ```train_loop``` inside ```train```? ```train``` is not readable now.

File: fluid/neural_machine_translation/transformer/train.py (hunk 1 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/train.py]: I move these functions out

File: fluid/neural_machine_translation/transformer/train.py (hunk 2 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/train.py]: a mistake, fix it

File: fluid/neural_machine_translation/transformer/README_cn.md [hunk 4]
```diff
@@ -154,9 +154,64 @@ perl multi-bleu.perl data/newstest2013.tok.de < predict.tok.txt
 ```
 BLEU = 25.08, 58.3/31.5/19.6/12.6 (BP=0.966, ratio=0.967, hyp_len=61321, ref_len=63412)
 ```
+###分布式训练
```
Comment (kuke) [inline fluid/neural_machine_translation/transformer/README_cn.md]: Please  pay attention to the format

File: fluid/neural_machine_translation/transformer/README_cn.md [hunk 5]
```diff
@@ -154,9 +154,64 @@ perl multi-bleu.perl data/newstest2013.tok.de < predict.tok.txt
...
+transformer 模型支持同步或者异步的分布式训练。分布式的配置主要两个方面:
+1 命令行配置
+  --local参数，这个参数有两个取值，True表示单机训练，而False表示使用分布式训练。默认为单机训练模式。
```
Comment (kuke) [inline fluid/neural_machine_translation/transformer/README_cn.md]: Please use `` for English names and values in text.

File: fluid/neural_machine_translation/transformer/train.py [hunk 6]
```diff
@@ -398,6 +404,7 @@ def test(exe=test_exe):
                 feed_dict[sum_cost.name + "@GRAD"] = 1. / total_num_token
             outs = train_exe.run(fetch_list=[sum_cost.name, token_num.name],
                                  feed=feed_list)
+            train_exe.bcast_params()
```
Comment (panyx0718) [inline fluid/neural_machine_translation/transformer/train.py]: 这个不需要了吧？

File: fluid/neural_machine_translation/transformer/README_cn.md (hunk 5 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/README_cn.md]: ok

File: fluid/neural_machine_translation/transformer/README_cn.md (hunk 4 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/README_cn.md]: ok

File: fluid/neural_machine_translation/transformer/train.py (hunk 3 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/train.py]: ok

File: fluid/neural_machine_translation/transformer/train.py (hunk 6 above)
Comment (xuezhong) [reply fluid/neural_machine_translation/transformer/train.py]: en  remove it

**Review by panyx0718 — APPROVED** [review]

**Comment by panyx0718** [comment]
@xuezhong 找 @kuke 确认一下吧

**Comment by kuke** [comment]
LGTM

**Pull request merged**